external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSinkSuite.scala

```diff
@@ -0,0 +1,413 @@
+        .option("kafka.metadata.max.age.ms", "1")
+        .option("startingOffsets", "earliest")
+        .option("subscribe", topic)
+        .option("failOnDataLoss", "true")
+        .load()
+      val kafka = reader
+        .selectExpr("CAST(key AS STRING)", "CAST(value AS STRING)")
+        .selectExpr("CAST(key AS INT)", "CAST(value AS INT)")
+        .as[(Int, Int)]
+
+      testStream(kafka, outputMode = OutputMode.Update)(
+        StartStream(ProcessingTime(0)),
+        AddMoreData(input, writer, "1", "2", "2", "3", "3", "3"),
+
+        CheckAnswer((1, 1), (2, 2), (3, 3)),
+        AddMoreData(input, writer, "1", "2", "3"),
+        CheckAnswer((1, 1), (2, 2), (3, 3), (1, 2), (2, 3), (3, 4)),
+        StopStream
+      )
+      writer.stop()
+    }
+  }
+
+  test("write structured streaming aggregation with topic field and topic option") {
+    /* The purpose of this test is to ensure that the topic option
+     * overrides the topic field. We begin by writing some data that
+     * includes a topic field and value (e.g., 'foo') along with a topic
+     * option. Then when we read from the topic specified in the option
+     * we should see the data i.e., the data was written to the topic
+     * option, and not to the topic in the data e.g., foo
+     */
+    withTempDir { checkpointDir =>
+      val input = MemoryStream[String]
+      val topic = newTopic()
+      testUtils.createTopic(topic)
+
+      val writer = input.toDF()
+        .groupBy("value")
+        .count()
+        .selectExpr("'foo' as topic",
+          "CAST(value as STRING) key", "CAST(count as STRING) value")
+        .writeStream
+        .format("kafka")
+        .option("checkpointLocation", checkpointDir.getCanonicalPath)
+        .outputMode(OutputMode.Update)
+        .option("kafka.bootstrap.servers", testUtils.brokerAddress)
+        .option("topic", topic)
+        .queryName("kafkaAggStream")
+        .start()
+
+      // Create Kafka source that reads from earliest to latest offset
+      val reader = spark.readStream
+        .format("kafka")
+        .option("kafka.bootstrap.servers", testUtils.brokerAddress)
+        .option("kafka.metadata.max.age.ms", "1")
+        .option("startingOffsets", "earliest")
+        .option("subscribe", topic)
+        .option("failOnDataLoss", "true")
+        .load()
+      val kafka = reader
+        .selectExpr("CAST(key AS STRING)", "CAST(value AS STRING)")
+        .selectExpr("CAST(key AS INT)", "CAST(value AS INT)")
+        .as[(Int, Int)]
+
+      testStream(kafka, outputMode = OutputMode.Update)(
+        StartStream(ProcessingTime(0)),
+        AddMoreData(input, writer, "1", "2", "2", "3", "3", "3"),
+
+        CheckAnswer((1, 1), (2, 2), (3, 3)),
+        AddMoreData(input, writer, "1", "2", "3"),
+        CheckAnswer((1, 1), (2, 2), (3, 3), (1, 2), (2, 3), (3, 4)),
+        StopStream
+      )
+      writer.stop()
+    }
+  }
+
+
+  test("write data with bad schema") {
```

Comment:
bad schema error should be thrown immediately when start() is called, rather later after the query has started. you have to validate before creating the sink object for this.